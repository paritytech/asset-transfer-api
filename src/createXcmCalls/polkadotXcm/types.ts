// Copyright 2023 Parity Technologies (UK) Ltd.

import type { XcmBaseArgs } from '../../types';

export type PolkadotXcmBaseArgs = XcmBaseArgs;
