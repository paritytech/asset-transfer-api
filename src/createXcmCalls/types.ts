// Copyright 2024 Parity Technologies (UK) Ltd.

export interface CreateXcmCallOpts {
	weightLimit?: { refTime?: string; proofSize?: string };
	paysWithFeeDest?: string;
	isLiquidTokenTransfer: boolean;
	isForeignAssetsTransfer: boolean;
	assetTransferType?: string;
	remoteReserveAssetTransferTypeLocation?: string;
	feesTransferType?: string;
	remoteReserveFeesTransferTypeLocation?: string;
	customXcmOnDest?: string;
}
