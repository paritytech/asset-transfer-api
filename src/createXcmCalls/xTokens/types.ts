import type { XcmBaseArgs } from '../../types';
import type { XcmPalletName } from '../util/establishXcmPallet';

export interface XTokensBaseArgs extends XcmBaseArgs {
	xcmPallet: XcmPalletName;
}
