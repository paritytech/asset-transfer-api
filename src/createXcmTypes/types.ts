// Copyright 2023 Parity Technologies (UK) Ltd.

import type { ApiPromise } from '@polkadot/api';
import type { AnyJson } from '@polkadot/types/types';

import type { Registry } from '../registry';
import type { RequireOnlyOne } from '../types';

export type InteriorValue = RequireOnlyOne<XcmJunctionDestBeneficiary> | XcmV4JunctionDestBeneficiary[] | null;

export type XcmDestBeneficiary = {
	[x: string]: {
		parents: number;
		interior: InteriorKey;
	};
};

export type InteriorKey = {
	[x: string]: InteriorValue;
};

export type XcmJunctionDestBeneficiary = {
	AccountId32: {
		network?: string;
		id: string;
	};
	AccountKey20: {
		network?: string;
		key: string;
	};
	Parachain: string;
	GlobalConsensus: string | AnyJson;
};

export type XcmV4JunctionDestBeneficiary =
	| {
			AccountId32: {
				network?: string;
				id: string;
			};
	  }
	| {
			Parachain: string;
	  }
	| {
			AccountKey20: {
				network?: string;
				key: string;
			};
	  }
	| {
			GlobalConsensus: string | AnyJson;
	  };

export type XcmV2MultiLocation = {
	parents: number;
	interior: RequireOnlyOne<XcmV2Junctions>;
};

export interface XcmV2Junctions {
	Here: '' | null;
	X1: XcmV2Junction;
	X2: [XcmV2Junction, XcmV2Junction];
	X3: [XcmV2Junction, XcmV2Junction, XcmV2Junction];
	X4: [XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction];
	X5: [XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction];
	X6: [XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction];
	X7: [XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction, XcmV2Junction];
	X8: [
		XcmV2Junction,
		XcmV2Junction,
		XcmV2Junction,
		XcmV2Junction,
		XcmV2Junction,
		XcmV2Junction,
		XcmV2Junction,
		XcmV2Junction,
	];
}

export type XcmV2Junction = RequireOnlyOne<XcmV2JunctionBase>;

export type XcmV2JunctionBase = {
	Parachain: number | string;
	AccountId32: { network?: XcmV2Network; id: string };
	AccountIndex64: { network?: XcmV2Network; id: string };
	AccountKey20: { network?: XcmV2Network; id: string };
	PalletInstance: number | string;
	GeneralIndex: string | number;
	GeneralKey: string;
	OnlyChild: AnyJson;
	Plurality: { id: AnyJson; part: AnyJson };
};

export type XcmV2Network = string | null;

export type XcmV3MultiLocation = {
	parents: number;
	interior: RequireOnlyOne<XcmV3Junctions>;
};

export interface XcmV3Junctions {
	Here: '' | null;
	X1: XcmV3Junction;
	X2: [XcmV3Junction, XcmV3Junction];
	X3: [XcmV3Junction, XcmV3Junction, XcmV3Junction];
	X4: [XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction];
	X5: [XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction];
	X6: [XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction];
	X7: [XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction, XcmV3Junction];
	X8: [
		XcmV3Junction,
		XcmV3Junction,
		XcmV3Junction,
		XcmV3Junction,
		XcmV3Junction,
		XcmV3Junction,
		XcmV3Junction,
		XcmV3Junction,
	];
}

export type XcmV3Junction = RequireOnlyOne<XcmV3JunctionBase>;

export type XcmV3JunctionBase = {
	Parachain: number;
	AccountId32: { network?: XcmV2Network; id: string };
	AccountIndex64: { network?: XcmV2Network; id: string };
	AccountKey20: { network?: XcmV2Network; id: string };
	PalletInstance: number;
	GeneralIndex: string | number;
	GeneralKey: string;
	OnlyChild: AnyJson;
	Plurality: { id: AnyJson; part: AnyJson };
	GlobalConsensus: string | AnyJson;
};

export type XcmV4Location = {
	parents: number;
	interior: RequireOnlyOne<XcmV4Junctions>;
};

export interface XcmV4Junctions {
	Here: '' | null;
	X1: [XcmV4Junction];
	X2: [XcmV4Junction, XcmV4Junction];
	X3: [XcmV4Junction, XcmV4Junction, XcmV4Junction];
	X4: [XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction];
	X5: [XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction];
	X6: [XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction];
	X7: [XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction, XcmV4Junction];
	X8: [
		XcmV4Junction,
		XcmV4Junction,
		XcmV4Junction,
		XcmV4Junction,
		XcmV4Junction,
		XcmV4Junction,
		XcmV4Junction,
		XcmV4Junction,
	];
}

export type XcmV4Junction = RequireOnlyOne<XcmV4JunctionBase>;

export type XcmV4JunctionBase = {
	Parachain: number;
	AccountId32: { network?: XcmV2Network; id: string };
	AccountIndex64: { network?: XcmV2Network; id: string };
	AccountKey20: { network?: XcmV2Network; id: string };
	PalletInstance: number;
	GeneralIndex: string | number;
	GeneralKey: string;
	OnlyChild: AnyJson;
	Plurality: { id: AnyJson; part: AnyJson };
	GlobalConsensus: string | AnyJson;
};

export type UnionJunction = XcmV2Junction | XcmV3Junction | XcmV4Junction;

export type UnionXcmMultiLocation = XcmV4Location | XcmV3MultiLocation | XcmV2MultiLocation;

export type UnionXcmMultiAssets = XcmV2MultiAssets | XcmV3MultiAssets | XcmV4Assets;

export type UnionXcmMultiAsset = XcmV2MultiAsset | XcmV3MultiAsset | XcmV4Asset;

export type UnionXcAssetsMultiAssets = XcAssetsV2MultiAssets | XcAssetsV3MultiAssets | XcAssetsV4MultiAssets;

export type UnionXcAssetsMultiAsset = XcAssetsV2MultiAsset | XcAssetsV3MultiAsset | XcAssetsV4Asset;

export interface XcmMultiAsset {
	id: {
		Concrete: UnionXcmMultiLocation;
	};
	fun: {
		Fungible: string;
	};
}

// XCM V4 Asset
export interface XcmAsset {
	id: UnionXcmMultiLocation;
	fun: {
		Fungible: string;
	};
}

export interface XcmV2MultiAssets {
	V2: XcmMultiAsset[];
}

export interface XcmV3MultiAssets {
	V3: XcmMultiAsset[];
}

export interface XcmV4Assets {
	V4: XcmAsset[];
}

export interface XcmV2MultiAsset {
	V2: XcmMultiAsset;
}

export interface XcmV3MultiAsset {
	V3: XcmMultiAsset;
}

export interface XcmV4Asset {
	V4: XcmAsset;
}

export interface XcAssetsV2MultiAssets {
	V2: FungibleObjMultiAsset[];
}

export interface XcAssetsV3MultiAssets {
	V3: FungibleObjMultiAsset[];
}

export interface XcAssetsV4MultiAssets {
	V4: FungibleObjAsset[];
}

export interface XcAssetsV2MultiAsset {
	V2: FungibleObjMultiAsset;
}

export interface XcAssetsV3MultiAsset {
	V3: FungibleObjMultiAsset;
}

export interface XcAssetsV4Asset {
	V4: FungibleObjAsset;
}

export type FungibleStrMultiAsset = {
	fun: {
		Fungible: string;
	};
	id: {
		Concrete: UnionXcmMultiLocation;
	};
};

export type FungibleStrAsset = {
	fun: {
		Fungible: string;
	};
	id: UnionXcmMultiLocation;
};

export type FungibleStrAssetType = FungibleStrMultiAsset | FungibleStrAsset;

export type FungibleObjMultiAsset = {
	fun: {
		Fungible: { Fungible: string };
	};
	id: {
		Concrete: UnionXcmMultiLocation;
	};
};

export type FungibleObjAsset = {
	fun: {
		Fungible: { Fungible: string };
	};
	id: UnionXcmMultiLocation;
};

export type FungibleObjAssetType = FungibleObjMultiAsset | FungibleObjAsset;

export type UnionXcAssetsMultiLocation = XcAssetsV2MultiLocation | XcAssetsV3MultiLocation | XcAssetsV4Location;

export interface XcAssetsV2MultiLocation {
	V2: {
		id: {
			Concrete: XcmV2MultiLocation;
		};
	};
}

export interface XcAssetsV3MultiLocation {
	V3: {
		id: {
			Concrete: XcmV3MultiLocation;
		};
	};
}

export interface XcAssetsV4Location {
	V4: {
		id: XcmV4Location;
	};
}

export interface XcmV2DestBeneficiary {
	V2: {
		parents: string | number;
		interior: {
			X1: { AccountId32: { id: string } };
		};
	};
}

export interface XcmV3DestBeneficiary {
	V3: {
		parents: string | number;
		interior: {
			X1: { AccountId32: { id: string } };
		};
	};
}

export interface XcmV4DestBeneficiary {
	V4: {
		parents: string | number;
		interior: {
			X1: [{ AccountId32: { id: string } }];
		};
	};
}

export interface XcmV2ParachainDestBeneficiary {
	V2: {
		parents: string | number;
		interior: {
			X2:
				| [{ Parachain: string }, { AccountId32: { id: string } }]
				| [{ Parachain: string }, { AccountKey20: { key: string } }];
		};
	};
}

export interface XcmV3ParachainDestBeneficiary {
	V3: {
		parents: string | number;
		interior: {
			X2:
				| [{ Parachain: string }, { AccountId32: { id: string } }]
				| [{ Parachain: string }, { AccountKey20: { key: string } }];
		};
	};
}

export interface XcmV4ParachainDestBeneficiary {
	V4: {
		parents: string | number;
		interior: {
			X2:
				| [{ Parachain: string }, { AccountId32: { id: string } }]
				| [{ Parachain: string }, { AccountKey20: { key: string } }];
		};
	};
}

export type XcmDestBeneficiaryXcAssets =
	| XcmV4DestBeneficiary
	| XcmV3DestBeneficiary
	| XcmV2DestBeneficiary
	| XcmV2ParachainDestBeneficiary
	| XcmV3ParachainDestBeneficiary
	| XcmV4ParachainDestBeneficiary;

export interface XcmWeightUnlimited {
	Unlimited: null | undefined;
}

export interface XcmWeightLimited {
	Limited: {
		refTime: string;
		proofSize: string;
	};
}

export type XcmWeight = XcmWeightUnlimited | XcmWeightLimited;

export interface CreateAssetsOpts {
	registry: Registry;
	isForeignAssetsTransfer: boolean;
	isLiquidTokenTransfer: boolean;
	api: ApiPromise;
}

export interface CreateFeeAssetItemOpts {
	registry: Registry;
	paysWithFeeDest?: string;
	specName?: string;
	assetIds?: string[];
	amounts?: string[];
	xcmVersion?: number;
	isForeignAssetsTransfer: boolean;
	isLiquidTokenTransfer: boolean;
}

export interface CreateWeightLimitOpts {
	weightLimit?: { refTime?: string; proofSize?: string };
}

export type XcmVersionedAssetIdV2 = {
	V2: {
		Concrete: UnionXcmMultiLocation;
	};
};
export type XcmVersionedAssetIdV3 = {
	V3: {
		Concrete: UnionXcmMultiLocation;
	};
};
export type XcmVersionedAssetIdV4 = {
	V4: UnionXcmMultiLocation;
};

export type XcmVersionedAssetId = XcmVersionedAssetIdV2 | XcmVersionedAssetIdV3 | XcmVersionedAssetIdV4;

export interface ICreateXcmType {
	createBeneficiary: (accountId: string, xcmVersion: number) => XcmDestBeneficiary;
	createDest: (destId: string, xcmVersion: number) => XcmDestBeneficiary;
	createAssets: (
		amounts: string[],
		xcmVersion: number,
		specName: string,
		assets: string[],
		opts: CreateAssetsOpts,
	) => Promise<UnionXcmMultiAssets>;
	createWeightLimit: (opts: CreateWeightLimitOpts) => XcmWeight;
	createFeeAssetItem: (api: ApiPromise, opts: CreateFeeAssetItemOpts) => Promise<number>;
	createXTokensBeneficiary?: (destChainId: string, accountId: string, xcmVersion: number) => XcmDestBeneficiaryXcAssets;
	createXTokensAssets?: (
		amounts: string[],
		xcmVersion: number,
		specName: string,
		assets: string[],
		opts: CreateAssetsOpts,
	) => Promise<UnionXcAssetsMultiAssets>;
	createXTokensAsset?: (
		amount: string,
		xcmVersion: number,
		specName: string,
		asset: string,
		opts: CreateAssetsOpts,
	) => Promise<UnionXcAssetsMultiAsset>;
	createXTokensWeightLimit?: (opts: CreateWeightLimitOpts) => XcmWeight;
	createXTokensFeeAssetItem?: (opts: CreateFeeAssetItemOpts) => UnionXcAssetsMultiLocation;
}
