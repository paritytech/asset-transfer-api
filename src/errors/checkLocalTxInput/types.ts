// Copyright 2023 Parity Technologies (UK) Ltd.

export enum LocalTxType {
	Assets = 'Assets',
	Balances = 'Balances',
	ForeignAssets = 'ForeignAssets',
	PoolAssets = 'PoolAssets',
	Tokens = 'Tokens',
}
