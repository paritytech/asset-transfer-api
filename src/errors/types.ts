// Copyright 2024 Parity Technologies (UK) Ltd.

export interface CheckXcmTxInputsOpts {
	isForeignAssetsTransfer: boolean;
	isLiquidTokenTransfer: boolean;
	isPrimaryParachainNativeAsset: boolean;
	paysWithFeeDest?: string;
	weightLimit?: { refTime?: string; proofSize?: string };
	assetTransferType?: string;
	remoteReserveAssetTransferTypeLocation?: string;
	feesTransferType?: string;
	remoteReserveFeesTransferTypeLocation?: string;
}
