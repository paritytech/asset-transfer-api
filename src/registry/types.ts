// Copyright 2023 Parity Technologies (UK) Ltd.

export interface AssetsInfo {
	[key: string]: string;
}

export interface ForeignAssetsData {
	symbol: string;
	name: string;
	multiLocation: string;
}

export interface ForeignAssetsInfo {
	[key: string]: ForeignAssetsData;
}

export type PoolPairsData = {
	[key: string]: {
		lpToken: string;
		pairInfo: string;
	};
};

export interface ChainInfoKeys {
	specName: string;
	tokens: string[];
	assetsInfo: AssetsInfo;
	foreignAssetsInfo: ForeignAssetsInfo;
	poolPairsInfo: PoolPairsData;
	xcAssetsData?: SanitizedXcAssetsData[];
}

export interface InjectedChainInfoKeys {
	specName?: string;
	tokens?: string[];
	assetsInfo?: AssetsInfo;
	foreignAssetsInfo?: ForeignAssetsInfo;
	poolPairsInfo?: PoolPairsData;
	xcAssetsData?: SanitizedXcAssetsData[];
}
export type ExpandedChainInfoKeys = { chainId: string } & ChainInfoKeys;

export type ChainInfo<T extends ChainInfoKeys | InjectedChainInfoKeys> = {
	[x: string]: T;
};

export type ChainInfoRegistry<T extends ChainInfoKeys | InjectedChainInfoKeys> = {
	polkadot: ChainInfo<T>;
	kusama: ChainInfo<T>;
	westend: ChainInfo<T>;
	rococo: ChainInfo<T>;
};

export type RelayChains = 'polkadot' | 'kusama' | 'westend' | 'rococo';

export type InterMultiLocationJunctionType = 'here' | 'x1' | 'x2' | 'x3' | 'x4' | 'x5' | 'x6' | 'x7' | 'x8';
type XCMRegistryInteriorMultiLocation = Partial<
	Record<InterMultiLocationJunctionType, null | XCMRegistryJunction | XCMRegistryJunctions>
>;

export type XCMRegistryJunction = {
	[x: string]: string | number | undefined | null | Partial<Record<string, string | number | undefined | null>>;
};
export type XCMRegistryJunctions = {
	[x: string]: string | number | undefined | null | Partial<Record<string, string | number | undefined | null>>;
}[];

export interface XCMRegistryMultiLocation {
	parents: number;
	interior: XCMRegistryInteriorMultiLocation;
}

export interface XCMAssetRegistryMultiLocation {
	v1: XCMRegistryMultiLocation;
}

export type SanitizedXcAssetsData = {
	paraID: number;
	symbol: string;
	decimals: number;
	xcmV1MultiLocation: string;
	asset:
		| { ForeignAsset: string }
		| { VToken: string }
		| { VToken2: string }
		| { VSToken: string }
		| { VSToken2: string }
		| { Token2: string }
		| { Token: string }
		| { Native: string }
		| { Stable: string }
		| string;
	assetHubReserveLocation: string;
	originChainReserveLocation?: string;
};

export type AssetsInfoType = 'assetsInfo' | 'foreignAssetsInfo' | 'poolPairsInfo';
