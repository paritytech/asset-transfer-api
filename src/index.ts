// Copyright 2024 Parity Technologies (UK) Ltd.

export * from './AssetTransferApi';
export * from './constructApiPromise';
export * from './types';
