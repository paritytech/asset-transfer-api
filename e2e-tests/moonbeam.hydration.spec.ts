import { setupNetworks, testingPairs } from '@acala-network/chopsticks-testing';
import { NetworkContext } from '@acala-network/chopsticks-utils';
import { AccountData } from '@polkadot/types/interfaces';
import { afterEach, beforeEach, expect, test } from 'vitest';

import { AssetTransferApi } from '../src/AssetTransferApi';

describe('Moonbeam <> Hydration', () => {
	let hydration: NetworkContext;
	let moonbeam: NetworkContext;
	const hydrationRecipientAddress = '15McF4S5ZsoAJGzdXE3FwSFVjSPoz1Cd7Xj7VQZCb7HULcjx';
	const hdxMoonbeamAssetID = '69606720909260275826784788104880799692';

	const { alice, alith } = testingPairs();

	beforeEach(async () => {
		const { hydration1, moonbeam1 } = await setupNetworks({
			hydration1: {
				endpoint: 'wss://rpc.hydradx.cloud',
				db: './db.sqlite',
				port: 8000,
			},
			moonbeam1: {
				endpoint: 'wss://moonbeam-rpc.dwellir.com',
				db: './db.sqlite',
				port: 8001,
			},
		});

		hydration = hydration1;
		moonbeam = moonbeam1;
	}, 500000);

	afterEach(async () => {
		await moonbeam.teardown();
		await hydration.teardown();
	}, 500000);

	test('Transfer GLMR from Moonbeam to Hydration', async () => {
		await moonbeam.dev.setStorage({
			System: {
				Account: [
					[[alith.address], { providers: 1, data: { free: '100000000000000000000000' } }], // GLMR
				],
			},
		});

		const recipientInitialGLMRBalance = await hydration.api.query.tokens.accounts(hydrationRecipientAddress, 16);
		expect((recipientInitialGLMRBalance as AccountData).free.toNumber()).toEqual(0);

		const assetTransferApi = new AssetTransferApi(moonbeam.api, 'moonbeam', 4);
		const tx = await assetTransferApi.createTransferTransaction(
			'2034',
			hydrationRecipientAddress,
			['GLMR'],
			['10000000000000000000'],
			{
				format: 'payload',
				xcmVersion: 4,
				sendersAddr: alith.address,
			},
		);

		const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

		await moonbeam.api.tx(extrinsic).signAndSend(alith);
		await moonbeam.dev.newBlock();

		await hydration.dev.newBlock();

		const recipientUpdatedGLMRBalance = await hydration.api.query.tokens.accounts(hydrationRecipientAddress, 16);

		expect((recipientUpdatedGLMRBalance as AccountData).free.toBigInt()).toBeGreaterThan(
			(recipientInitialGLMRBalance as AccountData).free.toBigInt(),
		);
	}, 100000);

	test('Transfer HDX from Hydration to Moonbeam', async () => {
		await hydration.dev.setStorage({
			System: {
				Account: [
					[[alice.address], { providers: 1, data: { free: '50000000000000000000000000' } }], // HDX
				],
			},
			Tokens: {
				Accounts: [
					[[alice.address, 0], { free: '50000000000000000000000000' }], // HDX
				],
			},
		});

		const recipientInitialHDXBalance = (
			await moonbeam.api.query.assets.account(hdxMoonbeamAssetID, alith.address)
		).unwrapOrDefault();
		expect(recipientInitialHDXBalance.balance.toNumber()).toEqual(0);

		const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', 4);
		const tx = await assetTransferApi.createTransferTransaction('2004', alith.address, ['HDX'], ['10000000000000000'], {
			format: 'payload',
			xcmVersion: 4,
			sendersAddr: alice.address,
		});

		const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

		await hydration.api.tx(extrinsic).signAndSend(alice);
		await hydration.dev.newBlock();

		await moonbeam.dev.newBlock();

		const recipientUpdatedHDXBalance = (
			await moonbeam.api.query.assets.account(hdxMoonbeamAssetID, alith.address)
		).unwrapOrDefault();
		expect(recipientUpdatedHDXBalance.balance.toBigInt()).toBeGreaterThan(
			recipientInitialHDXBalance.balance.toBigInt(),
		);
	}, 100000);
});
