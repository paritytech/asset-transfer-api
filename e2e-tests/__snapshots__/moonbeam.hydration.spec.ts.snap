// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Moonbeam <> Hydration > XCM V3 > Transfer GLMR from Moonbeam to Hydration > hydration recipients initial glmr balance 1`] = `
{
  "free": 0,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Moonbeam <> Hydration > XCM V3 > Transfer GLMR from Moonbeam to Hydration > hydration recipients updated glmr balance 1`] = `
{
  "free": "(rounded 10000000000000000000)",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`Moonbeam <> Hydration > XCM V3 > Transfer HDX from Hydration to Moonbeam > moonbeam recipients initial hdx balance 1`] = `
{
  "balance": 0,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`Moonbeam <> Hydration > XCM V3 > Transfer HDX from Hydration to Moonbeam > moonbeam recipients updated hdx balance 1`] = `
{
  "balance": "(rounded 10000000000000000)",
  "extra": null,
  "reason": {
    "sufficient": null,
  },
  "status": "Liquid",
}
`;
